zebra-state/src/service/write.rs

```diff
@@ -110,221 +113,341 @@ fn update_latest_chain_channels(
+struct WriteBlockWorkerTask {
+    finalized_block_write_receiver: UnboundedReceiver<QueuedCheckpointVerified>,
+    non_finalized_block_write_receiver: UnboundedReceiver<NonFinalizedWriteMessage>,
+    finalized_state: FinalizedState,
+    non_finalized_state: NonFinalizedState,
     invalid_block_reset_sender: UnboundedSender<block::Hash>,
-    mut chain_tip_sender: ChainTipSender,
+    chain_tip_sender: ChainTipSender,
     non_finalized_state_sender: watch::Sender<NonFinalizedState>,
-) {
-    let mut last_zebra_mined_log_height = None;
-    let mut prev_finalized_note_commitment_trees = None;
+}
+
+/// The message type for the non-finalized block write task channel.
+pub enum NonFinalizedWriteMessage {
+    /// A newly downloaded and semantically verified block prepared for
+    /// contextual validation and insertion into the non-finalized state.
+    Commit(QueuedSemanticallyVerified),
+    /// The hash of a block that should be invalidated and removed from
+    /// the non-finalized state, if present.
+    Invalidate(block::Hash),
+    /// The hash of a block that was previously invalidated but should be
+    /// reconsidered and reinserted into the non-finalized state.
+    Reconsider(block::Hash),
+}
+
+impl From<QueuedSemanticallyVerified> for NonFinalizedWriteMessage {
+    fn from(block: QueuedSemanticallyVerified) -> Self {
+        NonFinalizedWriteMessage::Commit(block)
+    }
+}
+
+/// A worker with a task that reads, validates, and writes blocks to the
+/// `finalized_state` or `non_finalized_state` and channels for sending
+/// it blocks.
+#[derive(Clone, Debug)]
+pub(super) struct BlockWriteSender {
+    /// A channel to send blocks to the `block_write_task`,
+    /// so they can be written to the [`NonFinalizedState`].
+    pub non_finalized: Option<tokio::sync::mpsc::UnboundedSender<NonFinalizedWriteMessage>>,
+
+    /// A channel to send blocks to the `block_write_task`,
+    /// so they can be written to the [`FinalizedState`].
+    ///
+    /// This sender is dropped after the state has finished sending all the checkpointed blocks,
+    /// and the lowest semantically verified block arrives.
+    pub finalized: Option<tokio::sync::mpsc::UnboundedSender<QueuedCheckpointVerified>>,
+}
+
+impl BlockWriteSender {
+    /// Creates a new [`BlockWriteSender`] with the given receivers and states.
+    #[instrument(
+        level = "debug",
+        skip_all,
+        fields(
+            network = %non_finalized_state.network
+        )
+    )]
+    pub fn spawn(
+        finalized_state: FinalizedState,
+        non_finalized_state: NonFinalizedState,
+        chain_tip_sender: ChainTipSender,
+        non_finalized_state_sender: watch::Sender<NonFinalizedState>,
+    ) -> (
+        Self,
+        tokio::sync::mpsc::UnboundedReceiver<block::Hash>,
+        Option<Arc<std::thread::JoinHandle<()>>>,
+    ) {
+        let (non_finalized_block_write_sender, non_finalized_block_write_receiver) =
```

Comment:
```suggestion
        // Security: The number of blocks in these channels is limited by
        //           the syncer and inbound lookahead limits.
        let (non_finalized_block_write_sender, non_finalized_block_write_receiver) =
```

Reply:
Applied in https://github.com/ZcashFoundation/zebra/pull/9266/commits/58bcfcd2a1509a35bd89194ceb29e9c10ba56eb1.